### DISCUSSION TITLE
Saving/flattening a pdf with form to pdf with just the data no widgets

### DISCUSSION THREAD

**Discussion by jonnya** [pymupdf/PyMuPDF#1650] [Q&A] [answered]
Hi Jorj -

I am trying to get a form pdf filled and saved. 
Looping through the widgets and updating.
Tried the code you suggested at the end.
The saved flattened pdf has the background image, logo etc but no data in it.
Not sure where to go with it.

import fitz

doc = fitz.open('form.pdf')
page = doc[0]

for field in page.widgets():
    print(field.field_label)
    print(field.field_name)
    field.value = "testing"
    field.update()
    print(field.value)

pdfbytes = doc.convert_to_pdf()
new = fitz.open("pdf", pdfbytes)
new.save("flattened.pdf")

**Comment by JorjMcKie** [comment]
I guess I would need an example. The following at least works fine. This the demo PDF:
[widgettest.pdf](https://github.com/pymupdf/PyMuPDF/files/8317964/widgettest.pdf)
![grafik](https://user-images.githubusercontent.com/8290722/159330857-084e2fc6-8c5b-4a77-91f4-c072908ae407.png)
Doing this:
```python
import fitz
doc=fitz.open("widgettest.pdf")
page=doc[0]
w=page.first_widget
w
'Text' widget on page 0 of widgettest.pdf
w.field_value
'Times-Roman'
w.field_value="some new text"
w.field_value
'some new text'
w.update()
pdfbytes=doc.convert_to_pdf()
new=fitz.open("pdf",pdfbytes)
new.save("x.pdf")
doc.close()
new.close()
new=fitz.open("x.pdf")
new.is_form_pdf  # the resulting PDF has no fields anymore!
False
```
leads to this:
![grafik](https://user-images.githubusercontent.com/8290722/159331118-1967b876-5ded-4bde-916c-f8de4f6fedf4.png)

**Comment by JorjMcKie** [comment, marked as answer]
Just realized it:
You must use `.field_value`, not just `.value`!

**Reply by jonnya** [reply]
Thats it !
Thank you * 100 !

**Reply by jonnya** [reply]
I have 1 more question on this.
The pdf is getting flattened and the fields show up with their strings fine.
There is another string that was a web link www.site.com and when flattened it collapsed the www. to 1 w in the same space as the s over typed. Any way to fix this?

**Reply by JorjMcKie** [reply]
Not really. The MuPDF function behind this is not bug free, and then there may be problems with one of the fonts around that part of the page.
It also may be some peculiarity in the way the original page is coded, etc.
